core/src/main/java/org/elasticsearch/cluster/routing/RoutingService.java

```diff
@@ -139,9 +166,14 @@ private void reroute() {
             if (lifecycle.stopped()) {
                 return;
             }
+            if (rerouting.compareAndSet(false, true) == false) {
```

Comment:
its the same thing we do in the GatewayAllocator, I find this one simpler to reason about be honest